File: agent/skills/telegram/cli/bubblelint.go
package main

// Bubble lint: catch wall-of-text chat sends before they reach the recipient.
// The "text in short bubbles, one thought per send" rule is enforced here at
// send time rather than left to the personality preset: a rule the agent must
// remember to apply is the weakest enforcement and regresses into blocks that
// read like an assistant, not a person. Linting the one place every send passes
// through makes it structural: a wall is rejected with the reason, so the agent
// re-sends as several short calls with its own break points. The single bypass
// is `--longform` (reference material the user asked for); `--message-file` sends are linted too.

import (
	"fmt"
	"regexp"
	"strings"
	"unicode/utf8"
)

// A bubble is "a few words to one line, rarely two" (personality SKILL.md).
const (
	bubbleMaxChars = 220           // a genuinely long single bubble
	bubbleSpace    = " \t\r\n\v\f" // the gap that makes a mark read as a full stop
)

var (
	bubbleURLRe        = regexp.MustCompile(`https?://\S+`)         // urls
	bubbleDecimalRe    = regexp.MustCompile(`\b\d+[.,]\d+\b`)       // decimals: 8.6, 86,5
	bubbleInitialismRe = regexp.MustCompile(`\b(?:[A-Za-z]\.){2,}`) // initialisms: W.A.S.T.E., U.K.
	// Abbreviations, an allowlist, so an unlisted one reads as a full stop. Every entry is a word
	// that is always followed by more, never one that can end a thought: protecting "etc." or "min."
	// would blank the stop in "eggs, milk, etc. also bread" and let the wall through. Dotted forms
	// (e.g., a.m., U.K.) are absent because bubbleInitialismRe already covers them.
	bubbleAbbrRe = regexp.MustCompile(`(?i)\b(?:mr|mrs|ms|dr|prof|jr|sr|st|vs|approx|fig` +
		`|jan|feb|mar|apr|jun|jul|aug|sept|sep|oct|nov|dec` +
		`|inc|ltd|co|corp|ave|rd|blvd|vol|ch|pp)\.`)
	bubbleEllipsisRe = regexp.MustCompile(`\.{3,}`) // ellipsis: a texting beat, not a full stop
	// A line-leading ordered-list marker ("1.", "2)") is not a full stop: each item is one
	// short thought. Only the marker is blanked, so a real wall inside an item still trips.
	bubbleListMarkerRe = regexp.MustCompile(`(?m)^[ \t]*\d+[.)][ \t]`)
	bubbleEnderRe      = regexp.MustCompile(`[.!?]+`)
)

// stripProtected blanks out spans whose '.', '?' or '!' are not full stops, so
// they cannot trip the lint.
func stripProtected(text string) string {
	for _, re := range []*regexp.Regexp{bubbleListMarkerRe, bubbleURLRe, bubbleDecimalRe, bubbleInitialismRe, bubbleAbbrRe, bubbleEllipsisRe} {
		text = re.ReplaceAllString(text, " ")
	}
	return text
}

// textAfterFullStop reports whether a '.', '!' or '?' has anything after it: the
// tell of a second thought crammed into the same bubble. A mark only reads as a
// full stop when whitespace follows it, so "main.py" and "example.com" stay
// single thoughts.
func textAfterFullStop(text string) bool {
	cleaned := strings.TrimSpace(stripProtected(text))
	for _, loc := range bubbleEnderRe.FindAllStringIndex(cleaned, -1) {
		rest := cleaned[loc[1]:]
		trimmed := strings.TrimLeft(rest, bubbleSpace)
		if trimmed == "" || len(trimmed) == len(rest) {
			continue // the mark ends the bubble, or no whitespace gap follows it
		}
		return true
	}
	return false
}

// bubbleLintReason returns a non-empty explanation when message is a wall (too
// many characters, or text carrying on past a full stop), or "" if it passes.
// The caller turns a non-empty reason into an error that blocks the send.
func bubbleLintReason(message string) string {
	nChars := utf8.RuneCountInString(message)
	var why []string
	if nChars > bubbleMaxChars {
		why = append(why, fmt.Sprintf("%d chars", nChars))
	}
	if textAfterFullStop(message) {
		why = append(why, "text after a full stop")
	}
	if len(why) == 0 {
		return ""
	}
	return "bubble lint: this send is a wall (" + strings.Join(why, ", ") + "). " +
		"texting rule: short bubbles, one thought per send, and don't use full stops at all. " +
		"split it into several separate send calls, a beat between each, one idea each. if this " +
		"is genuine reference material (a brief, a code block, a list they asked for), resend the " +
		"same command with --longform to bypass."
}

File: agent/core/skills/app-chat/cli/src/app_chat_cli/bubblelint.py
"""Bubble lint: reject wall-of-text chat sends before they reach the app.

Mirrors the telegram/whatsapp CLI bubble lint so app-chat enforces the same
"short bubbles, one thought per send" rule at send time. A rule the agent only
has to *remember* is the weakest enforcement and keeps regressing into one
multi-sentence block that reads like an assistant, not a person. Checking at
send time, where every send passes through, makes it structural: a wall is
rejected with the reason, so the agent re-sends as several short calls.

The single bypass is ``--longform``, for genuine reference material the user
asked for (a brief, a code block, a list).
"""

import re

# A bubble is "a few words to one line, rarely two" (personality SKILL.md).
BUBBLE_MAX_CHARS = 220  # a genuinely long single bubble

_URL_RE = re.compile(r"https?://\S+")  # urls
_DECIMAL_RE = re.compile(r"\b\d+[.,]\d+\b")  # decimals: 8.6, 86,5
_INITIALISM_RE = re.compile(r"\b(?:[A-Za-z]\.){2,}")  # initialisms: W.A.S.T.E., U.K.
_ABBR_RE = re.compile(
    r"\b(?:mr|mrs|ms|dr|prof|st|vs|etc|e\.g|i\.e|a\.m|p\.m|u\.k|u\.s|approx|no|fig)\.",
    re.IGNORECASE,
)
_ELLIPSIS_RE = re.compile(r"\.{3,}")  # ellipsis: a texting beat, not a full stop
_ENDER_RE = re.compile(r"[.!?]+")

_SPACE = " \t\r\n\v\f"


def _strip_protected(text: str) -> str:
    """Blank out spans whose '.', '?' or '!' are not sentence boundaries."""
    for rx in (_URL_RE, _DECIMAL_RE, _INITIALISM_RE, _ABBR_RE, _ELLIPSIS_RE):
        text = rx.sub(" ", text)
    return text


def text_after_full_stop(text: str) -> bool:
    """True when a '.', '!' or '?' has anything after it: the tell of a second
    thought crammed into the same bubble.

    A mark only reads as a full stop when whitespace follows it, so "main.py"
    and "example.com" stay single thoughts.
    """
    cleaned = _strip_protected(text).strip()
    for match in _ENDER_RE.finditer(cleaned):
        rest = cleaned[match.end() :]
        trimmed = rest.lstrip(_SPACE)
        if trimmed == "" or len(trimmed) == len(rest):
            continue  # the mark ends the bubble, or no whitespace gap follows it
        return True
    return False


def bubble_lint_reason(message: str) -> str:
    """Return a non-empty explanation when message is a wall (too many
    characters, or text carrying on past a full stop), or "" if it passes."""
    n_chars = len(message)
    why = []
    if n_chars > BUBBLE_MAX_CHARS:
        why.append(f"{n_chars} chars")
    if text_after_full_stop(message):
        why.append("text after a full stop")
    if not why:
        return ""
    return (
        "bubble lint: this send is a wall (" + ", ".join(why) + "). texting rule: short bubbles, one thought per send, and don't use "
        "full stops at all. split it into several separate send calls, a beat between "
        "each, one idea each. if this is genuine reference material (a brief, a code "
        "block, a list they asked for), resend the same command with --longform to "
        "bypass."
    )
